icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java

```diff
@@ -240,24 +294,93 @@ public static LongNameHandler forMeasureUnit(
                     }
                 }
             }
-            return forCompoundUnit(locale, unit, perUnit, width, rules, parent);
+            return forCompoundUnit(locale, unit, perUnit, width, unitDisplayCase, rules, parent);
         }
 
         String[] simpleFormats = new String[ARRAY_LENGTH];
-        getMeasureData(locale, unit, width, simpleFormats);
+        getMeasureData(locale, unit, width, unitDisplayCase, simpleFormats);
         // TODO(ICU4J): Reduce the number of object creations here?
         Map<StandardPlural, SimpleModifier> modifiers = new EnumMap<>(
                 StandardPlural.class);
         LongNameHandler result = new LongNameHandler(modifiers, rules, parent);
         result.simpleFormatsToModifiers(simpleFormats, NumberFormat.Field.MEASURE_UNIT);
+        if (simpleFormats[GENDER_INDEX] != null) {
+            result.gender = simpleFormats[GENDER_INDEX];
+        }
+
         return result;
     }
 
+    /**
+     * Loads and applies deriveComponent rules from CLDR's grammaticalFeatures.xml.
+     * <p>
+     * Consider a deriveComponent rule that looks like this:
+     *
+     * <deriveComponent feature="case" structure="per" value0="compound" value1="nominative"/>
+     * <p>
+     * Instantiating an instance as follows:
+     * <p>
+     * DerivedComponents d(loc, "case", "per", "foo");
+     * <p>
+     * Applying the rule in the XML element above, `d.value0()` will be "foo", and
+     * `d.value1()` will be "nominative".
+     * <p>
+     * In case of any kind of failure, value0() and value1() will simply return "".
+     */
+    private static class DerivedComponents {
+        /**
+         * Constructor.
+         * <p>
+         * The feature and structure parameters must be null-terminated. The string
+         * referenced by compoundValue must exist for longer than the
+         * DerivedComponents instance.
+         */
+        public DerivedComponents(ULocale locale,
+                                 String feature,
+                                 String structure,
+                                 String compoundValue) {
+            ICUResourceBundle derivationsBundle =
+                    (ICUResourceBundle) UResourceBundle
+                            .getBundleInstance(ICUData.ICU_BASE_NAME, "grammaticalFeatures");
+            derivationsBundle = (ICUResourceBundle) derivationsBundle.get("grammaticalData");
+            derivationsBundle = (ICUResourceBundle) derivationsBundle.get("derivations");
+
+            // TODO: use standard normal locale resolution algorithms rather than just grabbing language:
+            ICUResourceBundle stackBundle;
+            try {
+                stackBundle = (ICUResourceBundle) derivationsBundle.get(locale.getLanguage());
```

Comment:
```suggestion
            ICUResourceBundle stackBundle;
            try {
                // TODO: use standard normal locale resolution algorithms rather than just grabbing language:
                stackBundle = (ICUResourceBundle) derivationsBundle.get(locale.getLanguage());
```

Reply:
done